Kitodo/src/main/java/org/kitodo/production/services/file/FileService.java

```diff
@@ -1183,39 +1200,92 @@ private void addNewURIsToExistingPhysicalDivisions(Map<String, Map<Subfolder, UR
         }
     }
 
+    private void removeMissingMediaFromWorkpiece(List<String> mediaToRemove, Workpiece workpiece,
+                                                 Collection<Subfolder> subfolders) {
+        List<PhysicalDivision> pages = workpiece.getAllPhysicalDivisionChildrenFilteredByTypePageAndSorted();
+        for (String removal : mediaToRemove.stream().filter(Objects::nonNull).collect(Collectors.toList())) {
```

Comment:
Filtering `mediaToRemove` from `null` Strings looks like it is hiding a bug elsewhere. I assume, there shouldn’t be any `null` Strings in `mediaToRemove`. My suggestion would be not to check for it here at all.

If you absolutely **must** do this, don’t first make a list of the result again, for then only iterating over the list. Either:

    for (String removal : mediaToRemove) {
        if (Objects.isNull(removal)) {
            continue;
        }    

Or:

    mediaToRemove.stream().filter(Objects::nonNull).forEach(removal -> {

Reply:
Thanks for pointing this out. I changed this but did not just check just for `null`, but opted for a "blank" check instead to also filter emptry Strings here - which might be returned from `getCanonical` after implementing the proposed change by @henning-gerhardt. I hope that works for you, too.